### PR TITLE
dir, stored: start statistics threads only if needed

### PR DESCRIPTION
**Pull request by alaaeddineelamri**
#### Description:
The director statistics thread starts even when not needed. In order to optimize and have less unnecessary workload, the statistics thread for the director is not started by default, unless  `collect statistics` and `statistics collect interval` are both specified.

#### Please check

- [X] Short description and the purpose of this PR is present _above this paragraph_
- [X] Your name is present in the AUTHORS file (optional)

If you have any questions or problems, please give a comment in the PR.

### Helpful documentation and best practices

- [Git Workflow](https://docs.bareos.org/DeveloperGuide/gitworkflow.html)
- [Automatic Sourcecode Formatting](https://docs.bareos.org/DeveloperGuide/generaldevel.html#automatic-sourcecode-formatting)
- [Check your commit messages](https://docs.bareos.org/DeveloperGuide/gitworkflow.html#commits)

##### General

- [x] PR name is meaningful
- [x] Purpose of the PR is understood
- [x] Separate commit for this PR in the CHANGELOG.md, PR number referenced is same
- [x] Commit descriptions are understandable and well formatted

##### Source code quality

- [x] Source code changes are understandable
- [x] Variable and function names are meaningful
- [x] Code comments are correct (logically and spelling)
- [x] Required documentation changes are present and part of the PR
- [x] `bareos-check-sources --since-merge` does not report any problems
- [x] `git status` should not report modifications in the source tree after building and testing

### CODE REVIEW

File: core/src/dird/stats.cc [hunk 1]
```diff
@@ -322,7 +330,7 @@ int StartStatisticsThread(void)
 
   statistics_initialized = true;
 
-  return 0;
+  return 1;
```
Comment (pstorz) [inline core/src/dird/stats.cc:333]: Why do we have a return code here at all? All calls to "StartStatisticsThread()" that i saw never checked the return code. so can't we just change ther return code to void?

And can't we do the same with StartStatisticsThread() in sd_stats.cc also?

File: core/src/tests/addresses_and_ports.cc [hunk 2]
```diff
@@ -135,8 +85,8 @@ static bool test_sockets(int family, int port)
   bool result = true;
 
   if (family == 0) {
-    result = create_and_bind_v4socket(v4_fd, port)
-             && create_and_bind_v6socket(v6_fd, port);
+    result = create_and_bind_v4socket(v4_fd, port);
+    result &= create_and_bind_v6socket(v6_fd, port);
```
Comment (pstorz) [inline core/src/tests/addresses_and_ports.cc:89]: Why this change? I guess to avoid short-cicuit evalutiont, right? However, &= is a bitwise AND and not logical AND.
For clarification it might make sense to introduce `bool result_v4` and `bool result_v6`and then have `result = restult_v4 && result_v6`

**Review by pstorz — CHANGES_REQUESTED** [review]

File: core/src/dird/stats.cc (hunk 1 above)
Comment (pstorz) [reply core/src/dird/stats.cc:333]: In case the return code is required, we should check it in the code that calls StartStatisticsThread()

File: core/src/tests/addresses_and_ports.cc (hunk 2 above)
Comment (alaaeddineelamri) [reply core/src/tests/addresses_and_ports.cc:89]: good point

File: core/src/dird/stats.cc (hunk 1 above)
Comment (alaaeddineelamri) [reply core/src/dird/stats.cc:333]: You are right, the return value is not being used at all, but I believe we should make use of it. I already started using in the test, and I think we should even use the return value to send a message in case of a failed starting attempt. 
Let me know what you think of this

Comment (pstorz) [reply core/src/dird/stats.cc:333]: Ok, then the return code of course needs to be checked and handeled everywhere. So the question is what to do if the start fails? Do we only log that or do we even stop the daemon in that case?

**Review by pstorz — CHANGES_REQUESTED** [review]
The code that handles the return status of StartStatisticsThread() in both storage and director needs to be added.
Good work!

**Review by pstorz — APPROVED** [review]

File: core/src/dird/dird.cc [hunk 3]
```diff
@@ -469,7 +469,12 @@ int main(int argc, char* argv[])
 
   //   InitDeviceResources();
 
-  StartStatisticsThread();
+  if (!StartStatisticsThread()) {
+    BErrNo be;
+    Emsg1(M_WARNING, 0,
```
Comment (pstorz) [inline core/src/dird/dird.cc:474]:
```suggestion
    Emsg1(M_ERROR_TERM, 0,
```

File: core/src/dird/dird.cc [hunk 4]
```diff
@@ -673,7 +678,12 @@ bool DoReloadConfig()
     } else {  // no jobs running
       FreeSavedResources(&prev_config);
     }
-    StartStatisticsThread();
+    if (!StartStatisticsThread()) {
+      BErrNo be;
+      Emsg1(M_WARNING, 0,
```
Comment (pstorz) [inline core/src/dird/dird.cc:683]:
```suggestion
      Emsg1(M_ERROR_TERM, 0,
```

File: core/src/stored/stored.cc [hunk 5]
```diff
@@ -353,7 +353,12 @@ int main(int argc, char* argv[])
         me->jcr_watchdog_time); /* start JobControlRecord watchdogs etc. */
   }
 
-  StartStatisticsThread();
+  if (!StartStatisticsThread()) {
+    BErrNo be;
+    Emsg1(M_WARNING, 0,
```
Comment (pstorz) [inline core/src/stored/stored.cc:358]:
```suggestion
    Emsg1(M_ERROR_TERM, 0,
```

**Review by pstorz — CHANGES_REQUESTED** [review]
Good work, however I would use M_ERROR_TERM instead of M_WARNING if the statistics thread is required but cannot be started. This probably is a serious problem.

Other than that, the last commit message could be reformatted as it has a single "s" in one line that needs to be removed.

Thank you!

File: core/src/dird/dird.cc (hunk 3 above)
Comment (alaaeddineelamri) [reply core/src/dird/dird.cc]: This way, the daemons will always shutdown, since the default behavior is now to not start.
I changed the message to be displayed from within the `StartStatisticsThread()`

**Review by pstorz — APPROVED** [review]
Good work!
